src/ray/core_worker/common.h

```diff
@@ -20,6 +20,63 @@ struct RayFunction {
   const std::vector<std::string> function_descriptor;
 };
 
+/// Binary representation of ray object.
+class RayObject {
+ public:
+  /// Create a ray object instance.
+  ///
+  /// \param[in] data Data of the ray object.
+  /// \param[in] metadata Metadata of the ray object.
+  /// \param[in] copy_data Whether this class should hold a copy of data.
+  RayObject(const std::shared_ptr<Buffer> &data,
+            const std::shared_ptr<Buffer> &metadata = nullptr, bool copy_data = false)
+      : data_(data), metadata_(metadata), has_data_copy_(copy_data) {
+    if (has_data_copy_) {
+      // If this object is required to hold a copy of the data,
+      // make a copy if the passed in buffers don't already have a copy.
+      if (data_ && !data_->OwnsData()) {
+        data_ = std::make_shared<LocalMemoryBuffer>(data_->Data(), data_->Size(), true);
+      }
+
+      if (metadata_ && !metadata_->OwnsData()) {
+        metadata_ = std::make_shared<LocalMemoryBuffer>(metadata_->Data(),
+                                                        metadata_->Size(), true);
+      }
+    }
+
+    RAY_CHECK((data_ && data_->Size()) || (metadata_ && metadata_->Size()))
+        << "Data and metadat cannot be both empty.";
+  }
+
+  /// Return the data of the ray object.
+  const std::shared_ptr<Buffer> &GetData() const { return data_; };
+
+  /// Return the metadata of the ray object.
+  const std::shared_ptr<Buffer> &GetMetadata() const { return metadata_; };
+
+  uint64_t GetSize() const {
+    uint64_t size = 0;
+    size += (data_ != nullptr) ? data_->Size() : 0;
+    size += (metadata_ != nullptr) ? metadata_->Size() : 0;
+    return size;
+  }
+
+  /// Whether this object has metadata.
+  bool HasMetadata() const { return metadata_ != nullptr; }
```

Comment:
If the invariant is that the object has either `data_` or `metadata_`, please also add a `HasData` method.